src/ios/MultipeerConnectivity/THEMultipeerClient.m

```diff
@@ -0,0 +1,284 @@
+
+    THEMultipeerClientSession *session = (THEMultipeerClientSession *)p;
+
+    NSString *peerIdentifier = info[PEER_IDENTIFIER_KEY];
+    if (session && ([[session remotePeerID] hash] == [peerID hash]) && 
+        [peerIdentifier isEqualToString:[session remotePeerIdentifier]])
+    {
+      NSLog(@"client: Found existing peer: %@", peerIdentifier);
+
+      clientSession = session;
+      previouslyVisible = clientSession.visible;
+    }
+    else
+    {
+      NSLog(@"client: Found new peer: %@", peerIdentifier);
+
+      // We've found a new peer, create a new record
+      clientSession = [[THEMultipeerClientSession alloc] 
+                                    initWithLocalPeerID:_localPeerId 
+                                       withRemotePeerID:peerID 
+                               withRemotePeerIdentifier:peerIdentifier];
+    }
+
+    [clientSession setVisible:YES];
+    return clientSession;
+  }];
+
+  if (clientSession && previouslyVisible == NO)
+  {
+    // A new peer or one that has become visible again. Only
+    // contact delegate when the state changes (we get duplicates a lot)
+    if ([_multipeerSessionDelegate respondsToSelector:@selector(didFindPeerIdentifier:peerName:)])
+    {
+      [_multipeerSessionDelegate 
+        didFindPeerIdentifier:[clientSession remotePeerIdentifier] 
+                     peerName:info[PEER_NAME_KEY]];
+    }
+  }
+}
+
+// Notifies the delegate that a peer was lost.
+- (void)browser:(MCNearbyServiceBrowser *)browser
+       lostPeer:(MCPeerID *)peerID
+{
+  __block BOOL previouslyVisible = NO;
+  __block THEMultipeerClientSession *clientSession = nil;
+
+  // Update the peer's record under lock
+  [_clientSessions updateForPeerID:peerID
+                       updateBlock:^THEMultipeerPeerSession *(THEMultipeerPeerSession *p) {
+
+    clientSession = (THEMultipeerClientSession *)p;
+    assert([[clientSession remotePeerID] hash] == [peerID hash]);
+
+    previouslyVisible = clientSession.visible;
+
+    if (clientSession)
+    {
+      NSLog(@"client: lost peer: %@", [clientSession remotePeerIdentifier]);
+      // disconnect will clear up any networking resources we currently hold
+      [clientSession disconnect];
+      [clientSession setVisible:NO];
+    }
+
+    return clientSession;
+  }];
+    
+  if (clientSession)
+  {
+    if (previouslyVisible == YES)
+    {
+      // Let interested parties know we lost a peer, only do this on a state change
+      if ([_multipeerSessionDelegate respondsToSelector:@selector(didLosePeerIdentifier:)])
+      {
+        [_multipeerSessionDelegate didLosePeerIdentifier:[clientSession remotePeerIdentifier]];
+      }
+    }
+  }
+  else
+  {
+    // Probably shouldn't happen
+    NSLog(@"WARNING: lostPeer we didn't know about");
+  }
+}
+
+- (void)browser:(MCNearbyServiceBrowser *)browser didNotStartBrowsingForPeers:(NSError *)error
+{
+  NSLog(@"WARNING: didNotStartBrowsingForPeers");
```

Comment:
My tendency is to believe we should fire a network changed event when this occurs because it means that stuff isn't going to work. But I am worried about race conditions. Imagine an app gets this error but before the delegate it called the user calls stop and start and then things are o.k. and so we fire a network available event equals true. Then the thread that got this callback finally gets a chance to run and so fires its event even though it is actually outdated by this point. So now the node.js code thinks that the network is gone when in fact it is fine! Unless we can guarantee that the delegates are always run on the main thread aren't we going to have problems like this?